Container resolvers: add docs, typing and tests (#15614)
galaxyproject/galaxy: doc/source/admin/container_resolvers.rst

container resolver.
If all configured container resolvers failed, i.e. no container description
could be obtained, the tool is by default excuted using 
:doc:`standard dependency resolvers <dependency_resolvers>`, e.g. ``conda``.
Alternatively, if the excution environment specifies
`require_container <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L528>`_
the job fails in this case (TODO to be tested? or is it already).

Configuration:
--------------

The list of container resolvers is defined using YAML that can be
specified in an extra file (``container_resolvers_config_file``) or
inline the Galaxy configuration (``container_resolvers``). 
During the container resolution the resolvers specified in this
list are sequentially applied stopping at the first resolver that
yields a container description. 

The YAML defines a list of container resolvers and their properties.
This might look as follows:

.. code-block:: yaml

   - type: "cached_explicit_singularity"
   - type: "cached_mulled_singularity"
   - type: "mulled_singularity"
     auto_install: False

> would define a list of three container resolvers, and the last setting an option

Reviewer: This might be a good opportunity as well to add a sample file in YAML for this config file, that reflects how the same content of the XML would be written (sometimes mapping XML attributes and children to YAML is not always straightforward, specially if one lacks an understanding of the code loading this config). This could help in the transition from XML to YAML files for people.

Author: Indeed. 

And maybe also a schema definition? But I don't understand those yet. Guess it wont be more than a mapping anyway (except maybe for the mapping resolver).

Reviewer 2: +1 to YAML. Also, the container config is generally so simple that I'm putting it inline in galaxy.yml anyway.

are ignored.

.. note::

   Its's important to note that similarities in the names not necessarily
   imply any similarity in the function of the container resolvers.

Main resolver types:
--------------------

The main types of container resolvers follow this naming scheme: 
``[cached_][explicit,mulled][_singularity]``. That is

- a container resolver is either ``explicit`` or ``mulled``
- cached if it is prefixed with ``cached_`` and non-cached otherwise. 
- yield a container discription suitable for singularity if
  suffixed by ``_singularity`` and docker otherwise.

1. docker vs singularity

Galaxy can execute tools in containers using  ``docker`` or ``singularity``.
The corresponding container resolvers yield container descriptions suitable
for the corresponding "executor".
Note that for the execution with ``singularity`` Galaxy relies mostly on
docker containers that are either executed directly or are converted
to singularity images (except for explicit container requirements of
``type="singularity"``).

There are important differences between Galaxy's cached docker and singularity
container resolvers. The caching mechanism essentially executes a

> ``docker pull`` or ``singularity pull``, respectively. For docker this creates

Reviewer: Is the `singularity pull` currently being called? Last time I setup something relying on singularity, I had to have a mechanism of getting the tool containers apriori I think. Or maybe I did this to avoid issues with concurrent pulls when dataset collections where the first ones to request the pull of an image... not sure. Might be good to know as well if concurrent `singularity pull` calls are somehow avoided currently.

Author: Some of them do:

- `cached_explicit_singularity` triggers a pull over [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/explicit.py#L106)
- `mulled` and `mulled_singularity` trigger pulls [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L592) which are then executed in the respective methods [here for `mulled`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L540) and [here for `mulled_singularity`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L632)

The pull happens when `install=True` (which is apparently the case when called during job preparation and when build is triggered from the container dependencies tab in the admin UI) ... seems that the value of `auto_intstall` is important here as well since it determines if the container description uses the URI / the path to the image (but I don't fully understand this yet).

The `cached_mulled*` only lookup the cache.

an entry in the docker image cache (on the local node) whereas for
singularity an image file is created in the specified ``cache_directory``.
On distributed systems ``cache_directory`` needs to be accessible on all
compute nodes.

.. note::

   Using a cached docker resolver has no additional value on distributed compute
   systems since the cache is only available locally. 
   Therefore an additional ``docker inspect ... ; [ $? -ne 0 ] && docker pull ...``
   command is used in each job script.
   For distributed compute systems built in techniques of docker may be useful:
   https://docs.docker.com/registry/recipes/mirror/.

2. mulled vs explict

Mulled container resolvers apply for requirements defined by tools that are
a set of packages:

.. code-block:: xml

  <requirements>
      <requirement type="package" version="0.5">foo</requirement>
      <requirement type="package" version="1.0">bar</requirement>

>   </requirements>

Reviewer: Maybe a mention could be added here to the ability to create multi tool containers via the biocontainers UI / pull request mechanism? Or to subscribe your tool list to the CI that monitors tools that get added to the toolshed for this kind of tools?

Author: I wrote a bit here already https://github.com/galaxyproject/galaxy/blob/889f7f7413d0b3609b6829279f678e8a0e06e63f/doc/source/admin/container_resolvers.rst#building-resolver-types

It misses at the moment that also bioconda pushes to biocontainers. Maybe it should be moved to a separate subsection and also quay.org and namespaces need to be mentioned. Will put it on my todo list.

- ``cached_mulled``
- ``cached_mulled_singularity``

Furthermore there are the following excplit container resolvers:

- ``explicit``
- ``explicit_singularity``
- ``cached_explicit_singularity``

Note that there is no ``cached_explicit`` resolver.

Function of the ``resolve`` function of the main resolver types:
----------------------------------------------------------------

The resolve function is called when 

- opening the container tab in the dependency admin UI (with ``install=False``)
- triggering a build from the admin UI (``with install=True``)
- when a job is prepared (with ``install=True``)

If the ``resolve`` function implements the caching of images then this only
happens if ``install=True``.

1. Explicit resolvers

The uncached explicit resolvers (``explicit`` and ``explicit_singularity``) only
compute a container description using an URI that suites the ``docker`` or
``singularity``, resp.

The cached explicit resolver, i.e. ``cached_explicit_singularity`` (no docker
analogon available), will download the image to the ``cache_directory`` and
return a container description that points to the image file in the
``cache_directory``.

2. Mulled resolvers

> All mulled resolvers compute a mulled hash that desribes the reuirements and

Reviewer 3: ```suggestion
All mulled resolvers compute a mulled hash that describes the requirements and
```

Building resolver types:
------------------------

There are two container resolvers that locally create a mulled container.

- ``build_mulled``
- ``build_mulled_singularity``

Note that at the moment ``build_mulled_singularity`` requires docker for building.

Instead of using these, it might be better to create multi package containers
that are deployed to biocontainers using the infrastructure provided by the
`multi-package-containers <https://github.com/BioContainers/multi-package-containers>`_
repository.

This allready happens automatically for the tools in many tool repositories, see
`planemo monitor <https://github.com/galaxyproject/planemo-monitor>`_

Parameters:
-----------

- namespace
- hash_func
- shell

- auto_install TODO no idea what this is doing / what it is good for

- ``cache_directory``: defaults to ``container_image_cache_path`` set in galaxy.yml,
  i.e. ``"database/container_cache/"``. Applies to all singularity resolvers and sets
  the directory where to save images.
- ``cache_directory_cacher_type``: ``"uncached"`` (default) or ``"dir_mtime"``.
  The singularity resolvers iterate over the contents of the cache directory. The contents

>   of the directory can be accessed uncached (thn the file listing is computed for each access)

Reviewer 3: ```suggestion
  of the directory can be accessed uncached (in which case, the file listing is computed for each access)
```

If all configured container resolvers failed, i.e. no container description
could be obtained, the tool is by default executed using 
:doc:`standard dependency resolvers <dependency_resolvers>`, e.g. ``conda``.
Alternatively, if the execution environment specifies
`require_container <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L528>`_
the job fails in this case (TODO to be tested? or is it already).

Besides determining a container description some container resolvers
also cache/build containers.

Configuration:
--------------

The list of container resolvers is defined using YAML. This can be
either

- globally in an extra file (``container_resolvers_config_file``) or inline the Galaxy configuration (``container_resolvers``) or
- per execution environment using ``container_resolvers_config_file`` or ``container_resolvers``

Container resolvers defined for the execution environment
take precedence over globally defined container resolvers.
A sample YAML file showing the default configuration which is active
if neither a global or local configuration is given in

> `container_resolvers.yml.sample <https://github.com/galaxyproject/galaxy/tree/dev/lib/galaxy/config/sample/container_resolvers.yml.sample>`_.

Reviewer 4: Link not working I think.

Author: Should work when merged :)

.. _container_resolvers:

Containers in Galaxy
====================

> Galaxy can run tools using containers using ``docker`` or ``singularity``.

Reviewer 5: ```suggestion
Galaxy can run tools inside containers using ``docker`` or ``singularity``.
```

The containers can be either explicit or mulled (also called multi package containers).
The former are given by ``<container>`` requirements pointing to a specific container.

> The later are containers built for a set of requirements of type package.

Reviewer 5: ```suggestion
The latter are containers built for a set of requirements of type ``package``.
```

Mulled containers are described by a hash that is unique for a set of
packages and versions (for mulled v2), e.g. 
``mulled-v2-0d814cbcd5aa81b280ecadbee9e4aba8d9ab33f7:0fb38379c04f2a8a345a2c8f74b190ea9a51b6f3-0``
(mulled-v2-PACKAGEHASH:VERSIONHASH-BUILDNUMBER). For mulled containers
of single packages simply the package name and version are used instead of the hashes,
e.g. ``ucsc-liftover:357--h446ed27_4``.

> The bioconda and the Galaxy project provide infrastructure to create mulled

Reviewer 5: ```suggestion
Bioconda and the Galaxy project provide infrastructure to create mulled
```

containers and to make them globally available on the ``quay.io/biocontainers``
container registry.

1. For each bioconda package a container is deployed
2. Mulled containers are created and deployed by the infrastructure provided by the
   `multi-package-containers <https://github.com/BioContainers/multi-package-containers>`_
   repository. Mulled containers are added automatically to this repository for all tools
   in tool repositories that are crawled by the 
   `planemo monitor <https://github.com/galaxyproject/planemo-monitor>`_ repository

>    (which includes for instance IUC and many other tool repositories).

Reviewer 5: ```suggestion
   (which includes for instance tools-iuc and several other tool repositories).
```

`singularity_enabled <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L556>`_
enabled. Note, the links to the sample configurations exemplify this for local execution environments,
but this works for any environment as long as ``docker`` or ``singularity`` are
available.

For jobs that are sent to such an execution environment Galaxy tries to obtain a
container description by sequentially executing the configured container
resolvers (see below). The job is then executed using the description returned
by the first successful container resolver.
If all configured container resolvers failed, i.e. no container description
could be obtained, the tool is by default executed using 
:doc:`standard dependency resolvers <dependency_resolvers>`, e.g. ``conda``.
Alternatively, if the execution environment specifies
`require_container <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L528>`_
the job fails in this case.

> Besides determining a container description some container resolvers

Reviewer 5: ```suggestion
Besides determining a container description, some container resolvers
```

> also cache/build containers.

Reviewer 5: ```suggestion
also cache and/or build containers.
```

.. note::

   It's important to note that similarities in the names not necessarily
   imply any similarity in the function of the container resolvers.

There are the following mulled container resolvers:

- ``mulled``
- ``mulled_singularity``
- ``cached_mulled``
- ``cached_mulled_singularity``

Furthermore there are the following explicit container resolvers:

- ``explicit``
- ``explicit_singularity``
- ``cached_explicit_singularity``

Note that there is no ``cached_explicit`` resolver.

1. docker vs singularity
""""""""""""""""""""""""

Galaxy can execute tools in containers using ``docker`` or ``singularity``.
The corresponding container resolvers yield container descriptions suitable
for the corresponding "executor", i.e., docker (singularity, resp.)
container resolvers will resolve a container only in compute environments
with enabled docker (singularity, resp.). Thus, if only compute environments
with docker (resp. singularity) are present then singularity (resp. docker)
container resolvers are ignored (and may be omitted).

Note that, for the execution with ``singularity`` Galaxy relies mostly on
docker containers that are either executed directly or are converted

> to singularity images. An exception are for instance explicit container

Reviewer 5: ```suggestion
to singularity images. An exception is for instance explicit container
```

requirements of ``type="singularity"``.

2. mulled vs explicit
"""""""""""""""""""""

Mulled container resolvers apply for requirements defined by tools that are
a set of packages:

.. code-block:: xml

  <requirements>
      <requirement type="package" version="0.5">foo</requirement>
      <requirement type="package" version="1.0">bar</requirement>
  </requirements>

Explicit container resolvers apply for requirements defined by tools in the form
of a container requirement:

.. code-block:: xml

  <requirements>
      <container type="docker">quay.io/qiime2/core:2022.8</container>
  </requirements>

See also :ref:`additional_resolver_types`.

3. cached vs non-cached
"""""""""""""""""""""""

While non-cached resolvers will yield a container description pointing to an online

> available docker container cached resolvers will store container images on disk and

Reviewer 5: ```suggestion
available docker container, cached resolvers will store container images on disk and
```

use those. 

This distinction is the weakest: some (by name) non-cached container resolvers
can also resolve cached containers and are even responsible for the caching itself,
i.e. they execute a ``pull``.

There are important differences between Galaxy's cached docker and singularity
container resolvers. The caching mechanism essentially executes a
``docker pull`` or ``singularity pull``, respectively. For docker this creates
an entry in the docker image cache (on the local node) whereas for
singularity an image file is created in the specified ``cache_directory``.

> On distributed systems ``cache_directory`` needs to be accessible on all

Reviewer 6: but it doesn't need to be on shared storage, right ?

compute nodes.
For singularity admins should also take care of the ``APPTAINER_CACHEDIR``
directory.

.. note::

>    Using a cached docker resolver has no additional value on distributed compute

Reviewer 6: For starters the resolution is a lot faster, you should definitely not disable it, which one might think if they were to read this.

   systems since the cache is only available locally. 
   Therefore an additional ``docker inspect ... ; [ $? -ne 0 ] && docker pull ...``

>    command is used in each job script. Thereby a the container will be cached

Reviewer 5: ```suggestion
   command is used in each job script. Thereby a container will be cached
```

   after the tool run even if no cached container resolver was used.
   Clearly admins need to take care of docker caches of the main and compute nodes.

>    For distributed compute systems built in techniques of docker may be useful:

Reviewer 5: ```suggestion
   For distributed compute systems, built-in techniques of docker may be useful:
```

   https://docs.docker.com/registry/recipes/mirror/.

.. _function_of_the_resolve_function_of_the_main_resolver_types:

Function and use of the ``resolve`` function of the main resolver types:
------------------------------------------------------------------------

The resolve function is called when 

> 1. listing the the container tab in the dependency admin UI (using ``api/container_resolvers/toolbox``)

Reviewer 5: ```suggestion
1. listing the container tab in the dependency admin UI (using ``api/container_resolvers/toolbox``)
```

and returns a container description pointing to the cached image. For docker this is
done by executing ``docker images`` and for ``singularity`` the content of the
cache directory (``cache_directory``) is queried.

.. note::

    In contrast to the cached explicit resolver the cached mulled resolvers do not
    cache images, but they only query the available cached images.

The "uncached" mulled resolvers (``mulled`` and ``mulled_singularity``) by
default just return a container description containing the URI of the container
and download the image to the cache if ``install=True`` (see also
:ref:`function_of_the_resolve_function_of_the_main_resolver_types`). The caching
is done by a call to ``docker pull`` and ``singularity pull``, respectively.
Note that, by default the URI is returned in any case, i.e. even if the image
just has been downloaded or if the image is already in the cache. Only if the
resolvers are initialized with ``auto_install=True`` the ``resolve`` function
returns a container description pointing to the cached image. Note that this
makes a difference only for singularity (since for docker the URI is identical
to the name of the cached image).

.. note::

>     In contrast to the uncached explicit resolver the uncached mulled resolvers

Reviewer 5: ```suggestion
    In contrast to the uncached explicit resolver, the uncached mulled resolvers
```

    `multi-package-containers <https://github.com/BioContainers/multi-package-containers>`_
    repository, e.g. by adding more tool repositories to the
    `planemo monitor <https://github.com/galaxyproject/planemo-monitor>`_

Parameters:
-----------

- ``namespace`` defaults to ``"biocontainers"`` for the non-building and
  ``"local"`` for the building mulled resolvers. Available for all mulled
  container resolvers **except** ``cached_mulled_singularity``.
  Used to set the namespace that is used to query quay.io. Note that there
  is no `"local"` namespace at quay.io, but Galaxy uses it to refer
  to locally built images (that's why it is the default for the building
  resolvers).
- ``hash_func``: ``"v1"`` or ``"v2"`` (default: "v2"):
  Applies to all mulled container resolvers. Sets the version of the mulled
  hash that is used in the image name.
- ``shell`` Defaults to ``/bin/bash`` and sets the shell to be used in the container.
  Applies only to the resolvers listed in `Additional resolver types`_.
- ``auto_install``: defaults to ``True``. 
  Applies to ``mulled``, ``mulled_singularity``, ``build_mulled``, and ``build_mulled_singularity``.

>   For the non-building resolvers this controls if a contained description pointing to the

Reviewer 5: ```suggestion
  For the non-building resolvers this controls if a container description pointing to the
```

  cached image shall be returned (``auto_install==False``). For the building
  resolvers the parameter controls if the container should be built
  also if the resolve function is called with ``install=False`` (e.g. when listing
  the container in the Admin UI and no other container resolver worked for a tool).

.. note::

    Admins certainly should think carefully about ``auto_install``, since there are
    many scenarios where the default is not desirable.

- ``cache_directory``: applies to singularity container resolvers that allow to
  cache images and sets the directory where to save images.

>   If not set containers are saved in ``"database/container_cache/singularity/[explicit|mulled]"``.

Reviewer 5: ```suggestion
  If not set, containers are saved in ``"database/container_cache/singularity/[explicit|mulled]"``.
```

- ``cache_directory_cacher_type``: ``"uncached"`` (default) or ``"dir_mtime"``.
  The singularity resolvers iterate over the contents of the cache directory. The contents
  of the directory can be accessed uncached (in which case the file listing is computed for each access)
  or cached (then the listing is computed only if the mtime of the cache dir changes and on first access).
  (applies to all singularity resolvers that can cache images, except explicit_singularity)

> Note on the built in caching capabilities of singularity and docker

Reviewer 5: ```suggestion
Note on the built-in caching capabilities of singularity and docker
```

-------------------------------------------------------------------

> It is important to note that docker as well as singularity have their own builtin

Reviewer 5: ```suggestion
It is important to note that docker as well as singularity have their own built-in
```

caching mechanism.

> In case of docker a ``docker pull`` (e.g. executed from a container resolver) or

Reviewer 5: ```suggestion
In case of docker, a ``docker pull`` (e.g. executed from a container resolver) or
```

``docker run`` (e.g. executed on the compute node running the job) will add the
image to the **local** image cache.

> Galaxy's docker container resolvers rely on docker's built it image cache,

Reviewer 5: ```suggestion
Galaxy's docker container resolvers rely on docker's built-in image cache,
```

i.e. they query the image cache on the node that is executing Galaxy.
If the nodes that execute jobs are different from the node executing Galaxy
it's important to note that these nodes will have independent caches that
admins might want to control.

.. note::

   For the the execution of jobs Galaxy already implement the `support for using
   tarballs of container images
   <https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_classes.py#L319>`_.
   from ``container_image_cache_path`` (set in galaxy.yml) or the destination
   property ``docker_container_image_cache_path``. But at the moment none of the
   docker container resolvers creates these image tarballs.

Also singularity has its own caching mechanism and caches by default to ``$HOME/.singularity``.

> It may be cleaned regularly using ``singularity cache`` or be disabled by using the

Reviewer 5: ```suggestion
It can be cleaned regularly using the ``singularity cache`` command, or disabled by using the
```

> ``SINGULARITY_DISABLE_CACHE``. Environment variable.

Reviewer 5: ```suggestion
``SINGULARITY_DISABLE_CACHE`` environment variable.
```
